generate/src/contract/events.rs

```diff
@@ -247,13 +247,138 @@ fn expand_all_events(cx: &Context) -> TokenStream {
         return quote! {};
     }
 
+    let event_enum = expand_event_enum(cx);
+    let event_from_log = expand_event_from_log(cx);
+
     quote! {
         impl Contract {
             /// Returns a log stream with all events.
-            pub fn all_events(&self) -> self::ethcontract::dyns::DynAllEventsBuilder<
-                self::ethcontract::RawEventData,
-            > {
-                self.raw_instance().all_events()
+            pub fn all_events(&self) -> self::ethcontract::dyns::DynAllEventsBuilder<Event> {
+                self::ethcontract::dyns::DynAllEventsBuilder::new(
+                    self.raw_instance().web3(),
+                    self.address(),
+                )
+            }
+        }
+
+        #event_enum
+        #event_from_log
+    }
+}
+
+/// Expands into an enum with one variant for each distinct event type,
+/// including anonymous types.
+fn expand_event_enum(cx: &Context) -> TokenStream {
+    let variants = cx
+        .artifact
+        .abi
+        .events()
+        .map(|event| {
+            let struct_name = expand_struct_name(&event);
+            quote! {
+                #struct_name(self::events::#struct_name)
+            }
+        })
+        .collect::<Vec<_>>();
+
+    quote! {
+        /// A contract event.
+        pub enum Event {
+            #( #variants, )*
+        }
+    }
+}
+
+/// Expands the `TryFromLog` implementation for the event enum.
+fn expand_event_from_log(cx: &Context) -> TokenStream {
+    let all_events = {
+        let mut all_events = cx
+            .artifact
+            .abi
+            .events()
+            .map(|event| {
+                let struct_name = expand_struct_name(&event);
+
+                let name = Literal::string(&event.name);
+                let decode_event = quote! {
+                    log.decode(
+                        &Contract::artifact()
+                            .abi
+                            .event(#name)
+                            .expect("generated event decode")
+                    )
+                };
+
+                (event, struct_name, decode_event)
+            })
+            .collect::<Vec<_>>();
+
+        // NOTE: We sort the events by name so that the anonymous error decoding
+        //   is consistent. Since the events are stored in a `HashMap`, there is
+        //   no guaranteed order, and in the case where there is ambiguity in
+        //   decoding anonymous events, its nice if they follow some strict and
+        //   predictable order.
+        all_events.sort_unstable_by_key(|(event, _, _)| &event.name);
```

Comment:
You love that unstable sort! Is there any place it can't be used?

Reply:
Yes, when you want to preserve the original list order in the case the sorting keys are equal. Consider the following vector of objects:
```
let mut values = vec![
    Foo { bar: 3, baz: "message" },
    Foo { bar: 2, baz: "awesome" },
    Foo { bar: 2, baz: "second" },
];
```

An unstable sort by `bar`: `values.sort_unstable_by_key(|value| value.bar)`, yields either:
```
&[
    Foo { bar: 2, baz: "awesome" },
    Foo { bar: 2, baz: "second" },
    Foo { bar: 3, baz: "message" },
]
```
OR
```
&[
    Foo { bar: 2, baz: "second" },
    Foo { bar: 2, baz: "awesome" },
    Foo { bar: 3, baz: "message" },
]
```

And provides no guarantees on the resulting order when `bar == 2`, as there is more than one value in that case.

A stable sort: `values.sort_by_key(|value| value.bar)`, would always yields:
```
&[
    Foo { bar: 2, baz: "awesome" },
    Foo { bar: 2, baz: "second" },
    Foo { bar: 3, baz: "message" },
]
```

Stable sorts have the downside of needing additional allocations for the sort and being a little slower in the general case. From the [docs](https://doc.rust-lang.org/std/primitive.slice.html#method.sort):
> When applicable, unstable sorting is preferred because it is generally faster than stable sorting and it doesn't allocate auxiliary memory.